### PR TITLE
Add support for S3 template URL's and recursively track events from n…

### PR DESCRIPTION
**Pull request by deuscapturus**
…ested stacks (#1)

* add template url parameter, wip

* add context for errors.  Only pass template url or body on stack create, update and validate

* track events for nested stacks

* recursively get events from nested stacks

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,5 @@ chocolatey/forge/tools/VERIFICATION.txt
 chocolatey/**/*.exe
 chocolatey/**/*.nupkg
 chocolatey/**/*.zip
+cf-scripts
+forge
```
Comment (nathandines) [inline .gitignore:12]: Lines 11 and 12 here are fairly specific to your environment, I'd prefer if we didn't commit these to the repo

File: commands/deploy.go [hunk 2]
```diff
@@ -21,19 +22,34 @@ var deployCmd = &cobra.Command{
...
+		}
+
+		if templateFile != "" {
+			templateBody, err := ioutil.ReadFile(templateFile)
+			if err != nil {
+				log.Fatal(err)
+			}
+			stack.TemplateBody = string(templateBody)
+		}
+		if templateUrl != "" {
```
Comment (nathandines) [inline commands/deploy.go:50]: Please add a space in above here (between lines 49 and 50) to ensure consistent spacing between these conditional statements

File: commands/deploy.go [hunk 3]
```diff
@@ -21,19 +22,34 @@ var deployCmd = &cobra.Command{
 	Use:   "deploy",
 	Short: "Deploy a CloudFormation Stack",
 	Run: func(cmd *cobra.Command, args []string) {
-		// Read template-file
-		if templateFile == "" {
+		// Read template-file or use template-url
+
```
Comment (nathandines) [inline commands/deploy.go:26]: Please remove the space at line 26 to ensure consistent spacing of comments against the conditional statements

File: commands/deploy.go [hunk 4]
```diff
@@ -132,10 +149,17 @@ func init() {
 		"template-file",
 		"t",
 		"",
-		"Path to the CloudFormation template to be deployed",
+		"Path to the CloudFormation template to be deployed.  Must not specify template-url.",
```
Comment (nathandines) [inline commands/deploy.go:152]: Could we perhaps reword this to the following? 
```suggestion
		"Path to the CloudFormation template to be deployed (This argument conflicts with \"template-url\")",
```

File: commands/deploy.go [hunk 5]
```diff
@@ -132,10 +149,17 @@ func init() {
...
 	)
 	deployCmd.MarkFlagFilename("template-file")
 
+	deployCmd.PersistentFlags().StringVar(
+		&templateUrl,
+		"template-url",
+		"",
+		"S3 url to the CloudFormation template to be deployed.  Must not specify template-file.",
```
Comment (nathandines) [inline commands/deploy.go:160]: Could we perhaps reword this to the following? 
```suggestion
		"S3 URI of the CloudFormation template to be deployed (This argument conflicts with \"template-file\")",
```

File: forgelib/deploy.go [hunk 6]
```diff
@@ -18,13 +19,19 @@ type DeployOut struct {
 
 // Deploy will create or update the stack (depending on its current state)
 func (s *Stack) Deploy() (output DeployOut, err error) {
-	validationResult, err := cfnClient.ValidateTemplate(
-		&cloudformation.ValidateTemplateInput{
-			TemplateBody: aws.String(s.TemplateBody),
-		},
-	)
+	var validateConfig cloudformation.ValidateTemplateInput
```
Comment (nathandines) [inline forgelib/deploy.go:22]: Could we rename this variable from `validateConfig` to `templateValidationInput`? `validateConfig` could imply that we're validating some configuration using this?

File: forgelib/deploy.go [hunk 7]
```diff
@@ -91,28 +98,35 @@ TEMPLATE_PARAMETERS:
...
 		jsonStackPolicy, err := yaml.YAMLToJSON([]byte(s.StackPolicyBody))
 		if err != nil {
-			return output, err
+			return output, errors.Wrap(err, "Unable to convert template policy from YAML to JSON")
 		}
 		jsonStackPolicyString := string(jsonStackPolicy)
 		inputStackPolicy = &jsonStackPolicyString
 	}
 
+	createConfig := cloudformation.CreateStackInput{
```
Comment (nathandines) [inline forgelib/deploy.go:107]: Could we rename `createConfig` to `stackCreationInput` for the same reasons as `validateConfig` above?

File: forgelib/deploy.go [hunk 8]
```diff
@@ -127,28 +141,35 @@ TEMPLATE_PARAMETERS:
...
-				TemplateBody:    aws.String(s.TemplateBody),
-				Capabilities:    validationResult.Capabilities,
-				Tags:            tags,
-				Parameters:      inputParams,
-				RoleARN:         roleARN,
-				StackPolicyBody: inputStackPolicy,
-			},
-		)
+
+		updateConfig := cloudformation.UpdateStackInput{
```
Comment (nathandines) [inline forgelib/deploy.go:148]: Could we rename `updateConfig` to `stackUpdateInput` for the same reasons as `validateConfig` above?

File: forgelib/deploy.go [hunk 9]
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/pkg/errors"
```
Comment (nathandines) [inline forgelib/deploy.go:11]: I didn't know about this package. Seems pretty nice.

One thing though, I want to ensure that error handling across the application is consistent, so could we please remove this from this pull request, and create a separate issue or PR around adding context to all error handling across the application? This could perhaps be centralised this into a common function in this package which all everything could call when returning an error.

File: forgelib/events.go [hunk 10]
```diff
@@ -19,23 +19,45 @@ func (s *Stack) ListEvents(after *time.Time) (events []*cloudformation.StackEven
...
-	err = cfnClient.DescribeStackEventsPages(
-		&cloudformation.DescribeStackEventsInput{
-			StackName: &s.StackID,
-		}, func(page *cloudformation.DescribeStackEventsOutput, lastPage bool) bool {
-			for _, e := range page.StackEvents {
-				if e.Timestamp.UnixNano() > after.UnixNano() {
-					events = append(events, e)
-				}
+	// Find any nested stacks
+	allStacks := []string{s.StackID}
```
Comment (nathandines) [inline forgelib/events.go:23]: I feel like adding the stack discovery into here is making `ListEvents` do too much, and makes testing the code more challenging.

Could we break out the nested stack discovery into a separate function and add tests for it, have it called as part of `GetStackInfo` (so it only gets called once per stack) and perform a recursive call to `ListEvents`?

We might need to add a field of `NestedStacks` to the `Stack` struct which should be a slice of `Stack`s. The nested stacks should initialised as the type `Stack`. They would be populated in `GetStackInfo` like so:

```go
s.NestedStacks, err = s.getNestedStacks()
if err != nil {
	return err
}
```

I imagine that the resulting code in `ListEvents` may look something like this:

```go
for _, nestedStack := range s.NestedStacks {
	nestedStackEvents, err := nestedStack.ListEvents(after)
	if err != nil {
		return events, err
	}
	events = append(events, nestedStackEvents...)
}
```

There'd probably need to be a little work around the handling and initialisation of all of this, but I feel this might be a bit more testable, and would also reduce calls to `DescribeStackResources`

**Review by nathandines — CHANGES_REQUESTED** [review]
Hey @deuscapturus,

Thanks so much for taking the time to contribute to this

 I've added some comments around the code, most of them nit-picking, but I do have some further considerations around how the nested stacks would be queried.

Please see the comments inline.

**Comment by deuscapturus** [comment]
Thanks for making this tool.  It resolves a lot of problems with Cloudformation.  I'm trying to work quickly to get a POC done using your tool.  I submitted this PR too early.  Don't have time to take care of all of these requests.  I'll close this PR for now.  Feel free to pull in changes from my fork.
